oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/OAuth2AuthorizedClientProviderBuilder.java

```diff
@@ -0,0 +1,267 @@
+/*
+ * Copyright 2002-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.client;
+
+import org.springframework.security.oauth2.client.endpoint.OAuth2AccessTokenResponseClient;
+import org.springframework.security.oauth2.client.endpoint.OAuth2ClientCredentialsGrantRequest;
+import org.springframework.security.oauth2.client.endpoint.OAuth2RefreshTokenGrantRequest;
+import org.springframework.util.Assert;
+
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+/**
+ * A builder that builds a {@link DelegatingOAuth2AuthorizedClientProvider} composed of
+ * one or more {@link OAuth2AuthorizedClientProvider}(s) that implement specific authorization grants.
+ * The supported authorization grants are {@link #authorizationCode() authorization_code},
+ * {@link #refreshToken() refresh_token} and {@link #clientCredentials() client_credentials}.
+ * In addition to the standard authorization grants, an implementation of an extension grant
+ * may be supplied via {@link #provider(OAuth2AuthorizedClientProvider)}.
+ *
+ * @author Joe Grandja
+ * @since 5.2
+ * @see OAuth2AuthorizedClientProvider
+ * @see AuthorizationCodeOAuth2AuthorizedClientProvider
+ * @see RefreshTokenOAuth2AuthorizedClientProvider
+ * @see ClientCredentialsOAuth2AuthorizedClientProvider
+ * @see DelegatingOAuth2AuthorizedClientProvider
+ */
+public final class OAuth2AuthorizedClientProviderBuilder {
+	private final Map<Class<?>, Builder> builders = new HashMap<>();
+
+	private OAuth2AuthorizedClientProviderBuilder() {
+	}
+
+	/**
+	 * Returns a new {@link OAuth2AuthorizedClientProviderBuilder} for configuring the supported authorization grant(s).
+	 *
+	 * @return the {@link OAuth2AuthorizedClientProviderBuilder}
+	 */
+	public static OAuth2AuthorizedClientProviderBuilder withProvider() {
```

Comment:
I agree the name `OAuth2AuthorizedClientBuilder#builder()` is better and our convention.

> is this static factory in the right place

I externalized `OAuth2AuthorizedClientProviderBuilder` from the core API `OAuth2AuthorizedClientProvider` because `OAuth2AuthorizedClientProvider` should not have any dependencies or any knowledge on how to build the specific implementations, e.g. `AuthorizationCodeOAuth2AuthorizedClientProvider`, `RefreshTokenOAuth2AuthorizedClientProvider`, etc. Furthermore, user's may implement their own `OAuth2AuthorizedClientProvider` if they have custom grants which they would provide via `OAuth2AuthorizedClientProviderBuilder.withProvider().provider(OAuth2AuthorizedClientProvider)`.